lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -752,12 +770,68 @@ selectLatestCheckpoint wid = fmap entityVal <$>
 selectUTxO
     :: Checkpoint
     -> SqlPersistT IO [UTxO]
-selectUTxO (Checkpoint wid sl _parent _bh) = fmap entityVal <$>
+selectUTxO (Checkpoint wid _sl _parent _bh) = fmap entityVal <$>
     selectList
         [ UtxoWalletId ==. wid
-        , UtxoSlot ==. sl
+        , UtxoSlotSpent ==. Nothing
         ] []
 
+insertWhenNew
+    :: W.WalletId
+    -> UTxO
+    -> SqlPersistT IO ()
+insertWhenNew wid utxo@(UTxO _ _ _ inpId inpIx outAddr outAmt) = do
+    utxos <- selectList [ UtxoWalletId ==. wid
+                        , UtxoInputId ==. inpId
+                        , UtxoInputIndex ==. inpIx
+                        , UtxoOutputAddress ==. outAddr
+                        , UtxoOutputCoin ==. outAmt
+                        ] []
+    when (null utxos) $ insert_ utxo
+
+addSpentUtxos
+    :: [TxMeta]
+    -> [TxIn]
+    -> SqlPersistT IO ()
+addSpentUtxos metas txins = do
+    let outgoingAndInLedger =
+            filter (\(TxMeta _ _ st dir _ _) ->
+                        st == W.InLedger && dir == W.Outgoing) metas
+    mapM_ (updateUtxosPerMeta txins) outgoingAndInLedger
+
+updateUtxosPerMeta
+    :: [TxIn]
+    -> TxMeta
+    -> SqlPersistT IO ()
+updateUtxosPerMeta txins (TxMeta txid wid _ _ sl _) = do
+    let validTxIns = filter (\(TxIn txid' _ _ _ _ ) -> txid'==txid) txins
+    mapM_ (updateUtxoSpent wid sl) validTxIns
```

Comment:
We can be much simpler about these two functions above:

- For each transaction input, update if exists, all UTxO from our wallet for matching input id and input ix.

A UTxO is uniquely identified by its input id and input index. So there are two scenarios: 

- Either the UTxO is known and in the database, in which case, we have spent it.
- We don't know this (inputId, inputIx) and the update is without any effect

We should avoid putting some business logic inside the database layer. There's no interpretation of the data happening, no need to check for direction, no need to verify transaction ids :wink:

Reply:
Done